dotnet/efcore
# Working implementation of precompiled queries

Here's a first "semi-complete" working implementation of precompiled queries (#25009), for creating a feature branch. We can use this PR for first discussions.

To test this:

* Create a console program referencing this branch via `<ProjectReference>`, and run `dotnet ef dbcontext optimize` (I've replaced that to do precompiled querying, until we decide on the exact CLI switch (also to make it easier to test without needing a modified dotnet-ef tool)).
* Alternatively, you can do the precompiled query generation from your console app itself, which makes for a nice inner loop experience (you can debug into it etc.):

<details>
<summary>Triggering precompilation from your console app</summary>

```c#
if (args.Length > 0 && args[0] == "regenerate")
{
    var services = new ServiceCollection()
        .AddEntityFrameworkDesignTimeServices()
        .AddDbContextDesignTimeServices(ctx);
    var npgsqlDesignTimeServices = new SqlServerDesignTimeServices();
    npgsqlDesignTimeServices.ConfigureDesignTimeServices(services);
    var serviceProvider = services.BuildServiceProvider();
    var precompiledQueriesCodeGenerator = serviceProvider.GetRequiredService<IPrecompiledQueryCodeGenerator>();
    await precompiledQueriesCodeGenerator.GeneratePrecompiledQueries("/home/roji/projects/test/Test.csproj", ctx, outputDir: "/home/roji/projects/test");
    return;
}
```
</details>

* This builds the the project using Roslyn; note that this build is in additional to the regular build when invoking dotnet ef. Also, since you're referencing EF as a project reference, this will build EF too (long but only with this test setup).
* This reports the queries it identified, and generates an EFPrecompiledQueryBootstrapper into the project directory.
* When you run your console program, any query which was precompiled should be picked up, skipping runtime compilation.
* It can be useful to configure logging to show when query compilation and execution occur:

```c#
optionsBuilder.LogTo(Console.WriteLine, new[] { CoreEventId.QueryCompilationStarting, RelationalEventId.CommandExecuted })
```

</details>

This is by no means complete - I have a list of stuff that still needs to be done (split query, ExecuteUpdate/Delete, SQL queries, possibly some expression tree edge cases...).

To track issues you come across, I suggest posting them as a todo list in #25009  - no need for lots of different issues at this point).
